### PR TITLE
feat(delayed jobs): adding delayed jobs implementation

### PR DESCRIPTION
**Pull request by thoven87**
First pass at implementing delayed Jobs

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -10,7 +10,7 @@ let package = Package(
         .library(name: "JobsRedis", targets: ["JobsRedis"]),
     ],
     dependencies: [
-        .package(url: "https://github.com/hummingbird-project/swift-jobs.git", from: "1.0.0-beta.2"),
+        .package(url: "https://github.com/thoven87/swift-jobs.git", branch: "stevenson/delayed-jobs"),
```
Comment (Joannis) [inline Package.swift:13]: Marker: Change the version as appropriate, probably `beta.4` at that time

**Review by Joannis — APPROVED** [review]

**Review by adam-fowler — COMMENTED** [review]
So let me get this right
- "delayedZSet::JobId" stores the job buffer and a score (basically the delay in seconds)
- "delayedQueue" stores a queue of delayed job ids
Does this mean we have a zset for each queued job that is delayed, with one element in it? It appears you pull a job off the delayedQueue, then check if its associated sorted set score returns a value, if it doesn't you push it back onto the delayedQueue.

You could implement something similar by pushing jobID and delay into the standard queue, and when you pop off the standard queue compare the date with Date.now and if it is not ready push back on standard queue. This would remove the need for the additional sorted key and delayed queue.

Alternatively we could just have one sorted key which stores all the delayed job ids with a score based off the delay. Then each 10 or so ticks (this is not something we need to do on every call to popFirst) we can get the jobs off this queue whose score indicates they are no longer delayed and add them to the job queue.

The only problem with this is the read from sorted list and remove needs to be atomic. I can't find a redis command that does this. Maybe we could write a script to do that.

**Comment by thoven87** [comment]
> So let me get this right
> 
> * "delayedZSet::JobId" stores the job buffer and a score (basically the delay in seconds)
> * "delayedQueue" stores a queue of delayed job ids
>   Does this mean we have a zset for each queued job that is delayed, with one element in it? It appears you pull a job off the delayedQueue, then check if its associated sorted set score returns a value, if it doesn't you push it back onto the delayedQueue.

That's correct Adam! I could use `delayedZSet` which stores all job buffer and then move jobs out as needed. this could guarantee no data loss as no jobs are removed until `pool.zremrangebyscore` is called. 
> 
> You could implement something similar by pushing jobID and delay into the standard queue, and when you pop off the standard queue compare the date with Date.now and if it is not ready push back on standard queue. This would remove the need for the additional sorted key and delayed queue.

I think using the sorted key here is beneficial because it affords us to implement priority queue with less work. 
> 
> Alternatively we could just have one sorted key which stores all the delayed job ids with a score based off the delay. Then each 10 or so ticks (this is not something we need to do on every call to popFirst) we can get the jobs off this queue whose score indicates they are no longer delayed and add them to the job queue.
> 
> The only problem with this is the read from sorted list and remove needs to be atomic. I can't find a redis command that does this. Maybe we could write a script to do that.

Indeed, we could write a Lua script to do just that. I think a raised concerns about implementing such logic outside of swift. 

I was looking to see if there's a [watch](https://redis.io/docs/latest/commands/watch/) command for RedisStack. Maybe this could be used?

**Comment by thoven87** [comment]
> You could implement something similar by pushing jobID and delay into the standard queue, and when you pop off the standard queue compare the date with Date.now and if it is not ready push back on standard queue. This would remove the need for the additional sorted key and delayed queue.

This would require to have the delayUtil property set on the job instance. Moreover, the redis implementation would need to decode check date then encode if not read and push back no?

Could you please share what you had in mind?

**Comment by adam-fowler** [comment]
> This would require to have the delayUtil property set on the job instance. Moreover, the redis implementation would need to decode check date then encode if not read and push back no?

As with the Postgres and memory implementations the delayedUntil parameter only has to be stored in the queue. The queue currently only stores a job id.

**Comment by thoven87** [comment]
> > This would require to have the delayUtil property set on the job instance. Moreover, the redis implementation would need to decode check date then encode if not read and push back no?
> 
> As with the Postgres and memory implementations the delayedUntil parameter only has to be stored in the queue. The queue currently only stores a job id. Extending it to include the delayedUntil value wouldn't be hard.

Hmm, something like this `queue -> ["jobID::delayUntilValue"]`?

**Comment by adam-fowler** [comment]
The queue entries should be of the form
```
{"id":String,"delay":Int?}
```
Where id is the job id and delay is the delay until value as seconds from 1970.

**Comment by thoven87** [comment]
> The queue entries should be of the form
> 
> ```
> {"id":String,"delay":Int?}
> ```
> 
> Where id is the job id and delay is the delay until value as seconds from 1970.

I switched to using the delay job queue. I think we are all set since this version will handle migration just fine.

**Comment by thoven87** [comment]
What are the next steps to get this PR merged? @adam-fowler

**Comment by adam-fowler** [comment]
Let me have a look at it tomorrow. Sorry I haven't got to this. I've been quite busy

File: Sources/JobsRedis/RedisJobQueue.swift [hunk 2]
```diff
@@ -25,30 +26,43 @@ public final class RedisJobQueue: JobQueueDriver {
...
+            self.id = parts[0]
+            self.delayUntil = if parts.count > 0 {
+                Int(parts[1]) ?? 0
+            } else {
+                0
+            }
         }
 
         public init(from decoder: Decoder) throws {
             let container = try decoder.singleValueContainer()
```
Comment (adam-fowler) [inline Sources/JobsRedis/RedisJobQueue.swift:49]: Do we need the `init(from:)` and `encode(to:` anymore

**Review by adam-fowler — CHANGES_REQUESTED** [review]
In general looks goods, I've a couple of comments

File: Sources/JobsRedis/RedisJobQueue.swift (hunk 2 above)
Comment (thoven87) [reply Sources/JobsRedis/RedisJobQueue.swift:49]: no longer needed

File: Sources/JobsRedis/RedisJobQueue.swift [hunk 3]
```diff
@@ -94,9 +108,13 @@ public final class RedisJobQueue: JobQueueDriver {
     /// - Parameters:
     ///   - data: Job data
     /// - Returns: Queued job
-    @discardableResult public func push(_ buffer: ByteBuffer) async throws -> JobID {
-        let jobInstanceID = JobID()
+    @discardableResult public func push(_ buffer: ByteBuffer, options: JobOptions) async throws -> JobID {
+        let delay = options.delayUntil?.timeIntervalSince1970 ?? 0
+
+        let jobInstanceID = JobID(delayUntil: Int(delay))
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/JobsRedis/RedisJobQueue.swift]: I think we need to record this as an integer still, so multiple it by 1000 before converting it to a Int.

**Review by adam-fowler — CHANGES_REQUESTED** [review]
Delay still has to be recorded as an Int. You could also use the reference date (2000 I think) instead of 1970.

File: Sources/JobsRedis/RedisJobQueue.swift [hunk 4]
```diff
@@ -24,31 +25,47 @@ import RediStack
 public final class RedisJobQueue: JobQueueDriver {
     public struct JobID: Sendable, CustomStringConvertible {
         let id: String
+        let delayUntil: Int64
 
-        public init() {
+        public init(delayUntil: Double?) {
```
Comment (adam-fowler) [inline Sources/JobsRedis/RedisJobQueue.swift:30]: Why don't you hide the Date conversion completely and have the delayUntil value be a Date?
```suggestion
        public init(delayUntil: Date?) {
```

File: Sources/JobsRedis/RedisJobQueue.swift [hunk 5]
```diff
@@ -24,31 +25,47 @@ import RediStack
...
+                0
+            }
         }
 
-        public init(from decoder: Decoder) throws {
-            let container = try decoder.singleValueContainer()
-            self.id = try container.decode(String.self)
+        static func toMillis(value: Double?) -> Int64 {
+            if let value {
+                return Int64(value * 1000)
```
Comment (adam-fowler) [inline Sources/JobsRedis/RedisJobQueue.swift:49]: You check for overflow
```suggestion
                return Double(value * 1000) < Double(Int64.max) ? Int64(value * 1000) : Int64.max
```

**Pull request merged**